docs/proposals/rescheduling.md

```diff
@@ -0,0 +1,496 @@
+
+## Design proposal
+
+The key mechanisms and components of the proposed design are priority, preemption,
+disruption budgets, the `/evict` subresource, and the rescheduler.
+
+### Priority
+
+#### Motivation
+
+[Quality of Service levels](resource-qos.md) allow Kubernetes to safely overcommit nodes.
+Just as it is useful to overcommit nodes to increase node-level utilization, it is useful
+to overcommit clusters to increase cluster-level utilization.  Scheduling priority (which
+we abbreviate as *priority*, in combination with disruption budgets (described in the
+next section), allows Kubernetes to safely overcommit clusters much as QoS levels allow
+it to safely overcommit nodes.
+
+Today, cluster sharing among users, workload types, etc. is regulated via the
+[quota](../admin/resourcequota/README.md) mechanism. When allocating quota, a cluster
+administrator has two choices: (1) the sum of the quotas is less than or equal to the
+capacity of the cluster, or (2) the sum of the quotas is greater than the capacity of the
+cluster (that is, the cluster is overcommitted).  (1) is likely to lead to cluster
+under-utilization, while (2) is unsafe in the sense that someone's pods may go pending
+indefinitely even though they are still within their quota. Priority makes cluster
+overcommitment (i.e. case (2)) safe by allowing users and/or administrators to identify
+which pods should be allowed to run, and which should go pending, when demand for cluster
+resources exceeds supply to due to cluster overcommitment.
+
+Priority is also useful in some special-cause scenarios, such as ensuring that system
+DaemonSets can always schedule and reschedule onto every node where they want to run
+(assuming they are given the highest priority), e.g. see #21767.
+
+#### Specifying priorities
+
+We propose to add a required `Priority` field to `PodSpec`. Its value type is string, and
+the cluster administrator defines a total ordering on these strings (for example
+`Critical`, `Normal`, `Preemptible`). We choose string instead of integer so that it is
+easy for an administrator to add new priority levels in between existing levels, to
+encourage thinking about priority in terms of user intent and avoid magic numbers, and to
+make the internal implementation more flexible.
+
+When a scheduler is scheduling a new pod P and cannot find any node that meets all of P's
+scheduling predicates, it is allowed to evict ("preempt") one or more pods that are at
+the same or lower priority than P (subject to disruption budgets, see next section) from
+a node in order to make room for P, i.e. in order to make the scheduling predicates
+satisfied for P on that node.  (Note that when we add cluster-level resources (#19080),
+it might be necessary to preempt from multiple nodes, but that scenario is outside the
+scope of this document.)  The preempted pod(s) may or may not be able to reschedule. The
```

Comment:
@mqliang. I have taken a second thought on this. I think locking the schedulers at any time may not be a good idea, because pods that users submitted will get pending for quite a while as long as the rescheduler does not finish its job, which, unfortunately, in many use cases, is not acceptable: new job can not be started due to pod pending, application SLO may be violated due to failure of scaling up pods (newly created pods are scheduled by the scheduler(s)). The rescheduler's goal is to *OPTIMIZE* the cluster/certain-pod-placement, without harming user experience (or as less as possible). Note that the rescheduler is trying to improve the performance of the cluster/certain-pod: help the cluster/certain-pod to move from a "worth-improving" situation to a better one (locally), not from complete failure to "global optimization"; while locking the schedulers actually does "freeze" the whole cluster for quite some time, which is not worthy at all.

Regarding the global constraint-satisfaction problem brought up by @quinton-hoole , I don't think we could be able to achieve a truly "global optimized state", for:
1. it is too hard, if not impossible, to define the cost function. We need many a-prior knowledge of the system and the input. I believe there are many related research on the field;
2. in multi-scheduler scenario, it is even harder unless we introduce some cooperative constraints between the schedulers, which is also not easy;
3. I am very fond of the 3rd point brought by @quinton-hoole , that we can use some heuristics that depends heavily on the current pods placement. We can use such heuristics to see if there is anything we could do to improve the cluster/certain-pod, and the improvement can be measured by some cost function with a considerable punishment on moving the pods. +1 to
> (terminating and replacing pods creates a lot of churn in the system, so we'd better be sure that each of these gets us strictly closer to a near-optimal solution)

All in all, the above is just some of my thoughts and I believe there is no easy solution if we want the "global optimization". We should not consider this in the first step, or even for the next few steps. The rescheduler for the community version should stay simple, easy-to-understand, so let us leave the complexity to the researchers or dedicated commercial distributions.

FWIW, there is a project called [Firmament](www.firmament.io), in which the modeling can be a very good reference for those who are interested in the field. I am not sure if the authors are interested in this discussion. @ms705 @ICGog